src/utils/loadUmdBundle.ts

```diff
@@ -1,31 +1,86 @@
 import "@ungap/global-this";
 
-const umdBundlesPromiseCacheMap = new Map<string, Promise<unknown>>();
+import { retryFetch } from "./retryFetch";
+
+export interface LoadUmdBundleBackoff {
+  maxRetries: number;
+  delay: number;
+}
+
+interface UmdBundleCacheItem {
+  promise: Promise<unknown>;
+  timestamp: number;
+}
+
+const umdBundlesPromiseCacheMap = new Map<string, UmdBundleCacheItem>();
+
+export interface LoadUmdBundleOptions {
+  ttlInMs?: number;
+  backoff?: LoadUmdBundleBackoff;
+}
+
+const isCacheItemValid = ({
+  ttlInMs,
+  timestamp
+}: {
+  ttlInMs?: number;
+  timestamp: number;
+}) => !ttlInMs || Date.now() - timestamp < ttlInMs;
+
+export const loadUmdBundle = async <T>({
+  bundleUrl,
+  dependenciesMap,
+  loadUmdBundleOptions = {}
+}: {
+  bundleUrl: string;
+  dependenciesMap: Record<string, unknown>;
+  loadUmdBundleOptions: LoadUmdBundleOptions;
+}): Promise<T> => {
+  const { ttlInMs, backoff } = loadUmdBundleOptions;
 
-export const loadUmdBundle = async <T>(
-  bundleUrl: string,
-  dependenciesMap: Record<string, unknown>
-): Promise<T> => {
   if (umdBundlesPromiseCacheMap.has(bundleUrl)) {
-    return (await umdBundlesPromiseCacheMap.get(bundleUrl)) as Promise<T>;
+    const cacheItem = umdBundlesPromiseCacheMap.get(
+      bundleUrl
+    ) as UmdBundleCacheItem;
+
+    if (isCacheItemValid({ ttlInMs, timestamp: cacheItem.timestamp })) {
+      return (await cacheItem.promise) as Promise<T>;
+    }
   }
 
-  const umdBundlePromise = loadUmdBundleWithoutCache<T>(
+  const umdBundlePromise = loadUmdBundleWithoutCache<T>({
     bundleUrl,
-    dependenciesMap
-  );
-  umdBundlesPromiseCacheMap.set(bundleUrl, umdBundlePromise);
+    dependenciesMap,
+    backoff
+  });
+
+  umdBundlesPromiseCacheMap.set(bundleUrl, {
+    promise: umdBundlePromise,
+    timestamp: Date.now()
+  });
+
+  umdBundlePromise.catch(err => {
+    umdBundlesPromiseCacheMap.delete(bundleUrl);
+    throw err;
+  });
 
   return await umdBundlePromise;
 };
 
-export const loadUmdBundleWithoutCache = async <T>(
-  bundleUrl: string,
-  dependenciesMap: Record<string, unknown>
-): Promise<T> => {
-  const umdBundleSourceResponse = await fetch(bundleUrl);
+export const loadUmdBundleWithoutCache = async <T>({
+  bundleUrl,
+  dependenciesMap,
+  backoff
+}: {
+  bundleUrl: string;
+  dependenciesMap: Record<string, unknown>;
+  backoff?: LoadUmdBundleBackoff;
+}): Promise<T> => {
+  const umdBundleSourceResponse = backoff
+    ? await retryFetch({ loader: () => fetch(bundleUrl), options: backoff })
+    : await fetch(bundleUrl);
 
-  if (umdBundleSourceResponse.status >= 400) {
+  if (!umdBundleSourceResponse || umdBundleSourceResponse.status >= 400) {
```

Comment:
Not sure if `umdBundleSourceResponse == undefined` is ever possible 🤔, fetch should always throw or return a response.